### DISCUSSION TITLE
Preventing the implicit JOIN with super-tables

### DISCUSSION THREAD

**Discussion by anibali** [sqlalchemy/sqlalchemy#9765] [Usage Questions] [answered]
[In the documentation for writing SELECT statements for inheritance mappings](https://docs.sqlalchemy.org/en/20/orm/queryguide/inheritance.html#selecting-from-the-base-class-vs-specific-sub-classes), the following is stated:

> A SELECT statement constructed against a class in a joined inheritance hierarchy will query against the table to which the class is mapped, as well as any super-tables present, using JOIN to link them together.

Is there a way to prevent this JOIN and select from the specific sub-class table only when using `polymorphic_identity` and friends? I found a somewhat similar question on Stack Overflow but it is unanswered: https://stackoverflow.com/questions/49521591/sqlalchemy-how-to-select-attributes-from-a-polymorphic-child-without-joining-to.

To be more concrete, [following this example](https://docs.sqlalchemy.org/en/20/_modules/examples/inheritance/joined.html) how can I select an `Engineer` (columns id, status, engineer_name, primary_language) without also joining with the table for `Person` (columns company_id, name, type)?

**Comment by zzzeek** [comment, marked as answer]
you can do this using kind of a hack, which TBH either nobody has ever asked for before, or if it was asked for, it was over a decade ago before we had modern constructs that make this "possible" though note has not been tested for full robustness.

It's not as simple if the mapping has a polymorphic identity column since that has to be present in the result set for the ORM to know what type of object to load.   Even if you were loading Engineer objects, without the polymorphic identity column present, it would not know you are loading some sub-class of Engineer.

if you have a mapping like this:

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    adata: Mapped[str]


class B(A):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    bdata: Mapped[str]
```

you can query with this entity and get the thing you want:

```
BOnly = aliased(B, B.__table__)
```

that is, `select(BOnly)` or `session.query(BOnly)`.   that will give you really succinct queries against just the one table.  but if you have a polymorphic on, then you have to manufacture that column with a subquery.  given this mapping:


```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    adata: Mapped[str]
    type: Mapped[str]

    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_identity": "a"}

class B(A):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    bdata: Mapped[str]
    __mapper_args__ = {"polymorphic_identity": "b"}

```

we can query on just "B" but it needs to see "A.type" in there as well, we can use a fixed label and have it match like this:

```py
subq = select(B.__table__, literal("b").label("type")).subquery()
BOnly = aliased(B, subq, adapt_on_names=True)
```

that will query just "B" but queries will have a subquery in them, which might be less capable of optimization depending on your backend.    however you can query for B objects without accessing "A" and it seems to all check out, see POC below.

All of that said, if you are thinking you want this pattern,  which is not really idiomatic nor tested, consider not using inheritance and instead having the data that's in your "base" table accessed via a many-to-one relationship instead.   


```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select, literal
from sqlalchemy.orm import aliased
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    adata: Mapped[str]
    type: Mapped[str]

    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_identity": "a"}

class B(A):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    bdata: Mapped[str]
    __mapper_args__ = {"polymorphic_identity": "b"}


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(B(bdata="bd", adata="ad"))
s.commit()


subq = select(B.__table__, literal("b").label("type")).subquery()
BOnly = aliased(B, subq, adapt_on_names=True)

b1 = s.scalars(select(BOnly)).first()

# B attributes loaded
assert "bdata" in b1.__dict__

# A attributes did not
assert "adata" not in b1.__dict__

# attributes load as needed
assert b1.bdata == "bd"
assert b1.adata == "ad"

# load B normally
b2 = s.scalars(select(B)).first()

# load B using the base class
b3 = s.scalars(select(A)).first()

# assert these are all the same object in the identity map
assert b1 is b2
assert b1 is b3
```

**Reply by anibali** [reply]
Thank you for the extensive reply and the huge amount of work that you put into SQLAlchemy! I will try applying some of the approaches you've discussed here to my project and see how I go. If I can't achieve what I'm after I'll put together a minimal example that includes more of my other considerations, because although I've considered many different options for how to structure things and stared at the documentation a fair amount, I may have missed a more "idiomatic" solution.

**Comment by anibali** [comment]
Here's a more complete example of what I'm hoping to achieve.

```python
from sqlalchemy import ForeignKey, create_engine, UniqueConstraint, select, Table, Column, literal
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, Session, joinedload, with_polymorphic, relationship, \
    aliased

class Base(DeclarativeBase):
    pass

class Thing(Base):
    __tablename__ = 'thing'

    id: Mapped[int] = mapped_column(primary_key=True)
    type: Mapped[str]

    __mapper_args__ = {'polymorphic_on': 'type'}

class Book(Thing):
    __tablename__ = 'book'

    id: Mapped[int] = mapped_column(ForeignKey('thing.id'), primary_key=True)
    book_name: Mapped[str]

    __mapper_args__ = {'polymorphic_on': 'type', 'polymorphic_identity': 'book'}

# ...other sub-tables of Thing...

article_tag_table = Table(
    'article_tag',
    Base.metadata,
    Column('article_id', ForeignKey('article.id'), primary_key=True),
    Column('tag_id', ForeignKey('tag.id'), primary_key=True),
)

class Tag(Base):
    __tablename__ = 'tag'

    id: Mapped[int] = mapped_column(primary_key=True)
    tag_name: Mapped[str]
    # Tags are structured as a tree.
    parent_id: Mapped[int] = mapped_column(ForeignKey('tag.id'), nullable=True)
    # Tags can reference a polymorphic "thing" in the database.
    thing_id: Mapped[int] = mapped_column(ForeignKey('thing.id'), nullable=True)

    parent: Mapped['Tag'] = relationship('Tag', passive_deletes=True, foreign_keys=parent_id)
    thing: Mapped[Thing] = relationship('Thing')
    articles: Mapped[list['Article']] = relationship(secondary=article_tag_table, back_populates='tags')

    # Referenced "things" must be unique under a parent tag.
    __table_args__ = (
        UniqueConstraint('parent_id', 'thing_id', name='uniq_parent_thing'),
    )

class Article(Base):
    __tablename__ = 'article'

    id: Mapped[int] = mapped_column(primary_key=True)
    article_title: Mapped[str]

    tags: Mapped[list[Tag]] = relationship(secondary=article_tag_table, back_populates='articles')

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

#############

s = Session(e)

book_review_tag = Tag(tag_name='Book reviews')
book_promo_tag = Tag(tag_name='Book promotions')

book = Book(book_name='Moby Dick')
tag1 = Tag(tag_name='Moby Dick review', parent=book_review_tag, thing=book)
tag2 = Tag(tag_name='Moby Dick promo', parent=book_promo_tag, thing=book)

article = Article(article_title='An article about Moby Dick', tags=[tag1, tag2])

s.add(article)
s.commit()

#############

# Usage pattern 1: Retrieving specific "things" directly.
# I don't want to join with the Thing table here, since the super-table does not contain any useful information.
#
# Actual:  FROM thing JOIN book ON thing.id = book.id
# Desired: FROM book
books = s.scalars(select(Book)).all()
for book in books:
    print(book)

# This almost does what I want, except I'm worried about the performance impact of the subquery and I couldn't get it to
# work with an async session. Why do I have to round-trip the type through the query?
subq = select(Book.__table__, literal('book').label('type')).subquery()
BookOnly = aliased(Book, subq, adapt_on_names=True)
books = s.scalars(select(BookOnly)).all()
for book in books:
    print(book)

# Usage pattern 2: Polymorphism.
# I want to leverage SQLAlchemy's support for polymorphic tables to get tags along with associated "things".
PolymorphicThing = with_polymorphic(Thing, '*', flat=True)
tags_with_things = s.scalars(select(Tag).options(joinedload(Tag.thing.of_type(PolymorphicThing))))
for tag in tags_with_things:
    print(tag.tag_name, tag.thing)
```

To summarise:

1. I have a hierarchy (tree) of tags.
2. A tag can reference a "thing" in the database ("Book" is the example here).
3. Sibling tags can't reference the same "thing", and this is enforced with a `UniqueConstraint`.
4. I want to be able to get the "thing" for a tag using SQLAlchemy's polymorphic capabilities.
5. I want to be able to get a "thing" directly without joining the super-table.

An alternative that I've considered is making Tag polymorphic instead of Thing, which would result in super-table Tag and sub-table TagBook. Sub-tables could then have table-specific foreign keys (e.g. book_id instead of thing_id) and I wouldn't need the Thing table. The problem with that is that parent_id and book_id would be in two separate tables (Tag and TagBook respectively) so I couldn't add the UniqueConstraint. I need parent_id to be in a single table for all tags, not split across sub-tables, so moving parent_id to TagBook is not an option either. Hence I arrived at the structure I have outlined above in the code.

**Reply by zzzeek** [reply]
so this is kind of but maybe not exactly what we call a "generic association" and we have a lot of patterns for that in the examples at https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.generic_associations .      

So yes I agree this really looks like the Tag is what should be polymorphic.      The limitation is the desire for the unique constraint.

two ways to approach that are:

1. use single table inheritance for Tag / TagBook.  That way all the different types of "book_id", "other_thing_id" etc. columns are all on one table.  I noticed that your Tag.thing_id column is already nullable, so it would appear this is not an issue.
2. if TagBook is a separate table, duplicate the "parent_id" column on both Tag and TagBook; TagBook.parent_id would be a foreign key to Tag.parent_id so it's constrained in that direction, and TagBook.parent_id /TagBook.book_id would be within a unique constraint.   The ORM can write out the same value to this column on both tables using a multi-column property.

**Reply by anibali** [reply]
Excellent, you've given me plenty to go off of here and I'm confident that I can make something work (I'm liking the second approach at the moment). I know that you get a million questions, so once again I really appreciate you taking the time to help random internet strangers out!
